utils/status.go

```diff
@@ -0,0 +1,241 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/application-stacks/runtime-component-operator/common"
+	routev1 "github.com/openshift/api/route/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/types"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *ReconcilerBase) CheckApplicationStatus(ba common.BaseComponent) error {
+	// Create Application Ready status condition if it does not exit
+	s := ba.GetStatus()
+	conditionType := common.StatusConditionTypeReady
+	oldCondition := s.GetCondition(conditionType)
+
+	// Check resources status
+	r.CheckResourcesStatus(ba)
+
+	status, msg, reason := corev1.ConditionFalse, "", ""
+
+	// Check application and resources status
+	scReconciled := s.GetCondition(common.StatusConditionTypeReconciled)
+	scReady := s.GetCondition(common.StatusConditionTypeResourcesReady)
+
+	if scReconciled == nil || scReconciled.GetStatus() != corev1.ConditionTrue {
+		msg = "Application is not reconciled"
+		reason = "ApplicationNotReconciled"
+	} else if scReady == nil || scReady.GetStatus() != corev1.ConditionTrue {
+		msg = "Resources are not ready"
+		reason = "ResourcesNotReady"
+	} else {
+		status = corev1.ConditionTrue
+		msg = "Application is reconciled and resources are ready"
+	}
+
+	newCondition := s.NewCondition(conditionType)
+	newCondition.SetConditionFields(msg, reason, status)
+
+	return r.updateStatusConditions(ba, oldCondition, newCondition)
+}
+
+func (r *ReconcilerBase) CheckResourcesStatus(ba common.BaseComponent) error {
+	// Create Resource Ready status condition if it does not exit
+	s := ba.GetStatus()
+	conditionType := common.StatusConditionTypeResourcesReady
+	oldCondition := s.GetCondition(conditionType)
+	newCondition := s.NewCondition(conditionType)
+
+	// Check for Deployment, StatefulSet replicas or Knative service status
+	if ba.GetCreateKnativeService() == nil || !*ba.GetCreateKnativeService() {
+		newCondition = r.areReplicasReady(ba, newCondition)
+	} else {
+		newCondition = r.isKnativeReady(ba, newCondition)
+	}
+
+	return r.updateStatusConditions(ba, oldCondition, newCondition)
+}
+
+func (r *ReconcilerBase) updateStatusConditions(ba common.BaseComponent, oldCondition common.StatusCondition, newCondition common.StatusCondition) error {
```

Comment:
Suggest to split this method into two. One to set condition. Other to update status. Call the update status only once at the end of CheckApplicationStatus. That'll minimize the number of times status is actually updated on the cluster.